Adding OpenMC SS activation simulation

anu1217 opened the pull request:
fixes #5
- Adding OpenMC simulation
- Including .h5 files and nuclide density plot

anu1217 committed:
Adding ALARA input code, results & data files

anu1217 committed:
Adding OpenMC SS simulation & associated file

gonuke commented on fendl2bin.lib:
There is no need to put the data library in the repo

gonuke commented on myElelib:
Shouldn't need to put this element library in the repo if it's the standard one

gonuke commented on statepoint.10.h5:
We should discuss which output files should be in the repo, and if so, where they should be located

gonuke commented on tallies.out:
This file is probably not useful for the repo

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+import openmc
+import openmc.deplete
+import numpy as np
+from fractions import Fraction
```
do you use this?

anu1217 replied:
I was using this because of the 4/3 factor that appears in the spherical volume of depletable material but switched to 1.33 at some point. I now see that the Fraction module isn't necessary even when 4/3 is used -- getting rid of it

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+# An energy filter is created to assign to the flux tally.
+e_min_1, e_max_1 = 5e2, 14.001e6
+groups = 500
+energies = np.logspace(np.log10(e_min_1), np.log10(e_max_1), groups + 1)
```
Use vitamin-J group structure here.

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+e_min_2, e_max_2 = 5e2, 14.001e6
+groups = 500
+energies_2 = np.logspace(np.log10(e_min_2), np.log10(e_max_2), groups + 1)
+energy_filter_ng = openmc.EnergyFilter(energies_2)
```
This energy group structure is the same as `energies` - you don't need a different one, use the same one twice

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+energies_2 = np.logspace(np.log10(e_min_2), np.log10(e_max_2), groups + 1)
+energy_filter_ng = openmc.EnergyFilter(energies_2)
+
+spectrum_tally = openmc.Tally(name="Flux spectrum")
+# Implementing energy and cell filters for flux spectrum tally
+spectrum_tally.filters = [energy_filter_flux, cell_filter]
+spectrum_tally.scores = ['flux']
```
Can't you just apply these filters to the `neutron_tally`?

anu1217 replied:
I think I was looking for a neutron flux as a function of energy and just a total elastic/absorption score.

gonuke replied:
Does OpenMC give you the total over energy in addition to the spectrum when you apply an energy filter?

anu1217 replied:
Yes, it can provide a total tally in addition to the spectrum, but both have to be asked for explicitly. The total values should be the same as in PR #3 , but the spectrum may be different as we're accounting for Vitamin-J now

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+model = openmc.model.Model(geometry=geometry,settings=settings)
+#Depletion calculation
+W.depletable = True
+W.volume = 1.33333 * np.pi * (R_2**3 - R_1**3) #volume of W wall material
```
```suggestion
W.volume = 4.0/3.0 * np.pi * (R_2**3 - R_1**3) #volume of W wall material
```

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,166 @@
+#Depletion calculation
+W.depletable = True
+W.volume = 1.33333 * np.pi * (R_2**3 - R_1**3) #volume of W wall material
+chain = openmc.deplete.chain.Chain.from_xml("./chain_endfb71_sfr.xml")
```
This file name was assigned to a variable above - use that variable here to assure consistency... and below

gonuke commented on Nuclide_density_OpenMC.png:
might need this on a log plot to see things properly - note that 5 of these are stable and won't change as a function of the shutdown time

gonuke reviewed (changes requested):
This is a relatively short irradiation, with a total source of 1e20 n in a 10m radius sphere.  The uncollided flux arriving at the sphere is just under 8E10 with only 100 s of irradiation.  I would expect very little activation to occur!  You probably want multiple years of irradiation, e.g. 3e8 s for nearly 10 years.

anu1217 committed:
Update SS_Activation_OpenMC.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Removing 'import fraction' line at the top since module is not used

anu1217 committed:
Removed redundant line defining chain file + changed operating/cooling times

- Since the path for the chain file is already specified at the very top, it does not need to reappear in the operator

- Changed time_steps to 3e8 s of operation, followed by 24h and 30d of cooling

anu1217 committed:
Adding Vitamin J bounds in .txt format

anu1217 committed:
Deleted redundant energy bounds + used Vit-J bounds as flux bins

- Added filepath to text file with Vitamin-J energy bounds

- Redefined flux bins to follow Vitamin-J

anu1217 committed:
Generating semilog plot of nuclide density

- Making a semilog plot of density for readability
- Commenting out densities of long-lived isotopes

anu1217 committed:
Modifying accuracy of energy bins to 4th decimal point

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,150 @@
+time, n_W180 = results.get_atoms('1', 'W180', nuc_units = 'atom/cm3')
+time, n_W181 = results.get_atoms('1', 'W181', nuc_units = 'atom/cm3')
+time, n_W182 = results.get_atoms('1', 'W182', nuc_units = 'atom/cm3')
+time, n_W183 = results.get_atoms('1', 'W183', nuc_units = 'atom/cm3')
+time, n_W184 = results.get_atoms('1', 'W184', nuc_units = 'atom/cm3')
+time, n_W185 = results.get_atoms('1', 'W185', nuc_units = 'atom/cm3')
+time, n_W186 = results.get_atoms('1', 'W186', nuc_units = 'atom/cm3')
+time, n_W187 = results.get_atoms('1', 'W187', nuc_units = 'atom/cm3')
+time, n_Re185 = results.get_atoms('1', 'Re185', nuc_units = 'atom/cm3')
+time, n_Re187 = results.get_atoms('1', 'Re187', nuc_units = 'atom/cm3')
+time, n_Os187 = results.get_atoms('1', 'Os187', nuc_units = 'atom/cm3')
+
+print(time, n_He4)
+print(time, n_W179)
+print(time, n_W180)
+print(time, n_W181)
+print(time, n_W182)
+print(time, n_W183)
+print(time, n_W184)
+print(time, n_W185)
+print(time, n_W186)
+print(time, n_W187)
+print(time, n_Re185)
+print(time, n_Re187)
+print(time, n_Os187)
+
+# Plotting the number densities of each nuclide:
+plt.plot(time, n_He4, marker='.', linestyle='solid', color='red', label='He4')
+plt.plot(time, n_W179, marker='.', linestyle='solid', color='blue', label="W179")
+#plt.plot(time, n_W180, marker='.', linestyle='solid', color='green', label="W180")
+#plt.plot(time, n_W181, marker='.', linestyle='solid', color='deepskyblue', label="W181")
+#plt.plot(time, n_W182, marker='.', linestyle='solid', color='peru', label="W182")
+#plt.plot(time, n_W183, marker='.', linestyle='solid', color='plum', label="W183")
+#plt.plot(time, n_W184, marker='.', linestyle='solid', color='crimson', label="W184")
+plt.plot(time, n_W185, marker='.', linestyle='solid', color='rosybrown', label="W185")
+#plt.plot(time, n_W186, marker='.', linestyle='solid', color='gray', label="W186")
+plt.plot(time, n_W187, marker='.', linestyle='solid', color='lime', label="W187")
+plt.plot(time, n_Re185, marker='.', linestyle='solid', color='gold', label="Re185")
+plt.plot(time, n_Re187, marker='.', linestyle='solid', color='tomato', label="Re187")
+plt.plot(time, n_Os187, marker='.', linestyle='solid', color='cyan', label="Os187")
```
Some kind of list data structure may be better?
```suggestion
num_dens = {}
plot_colors = {'He4' : 'red', 'W179' : 'green', 'W180' : blue } #etcetera

for nuclide, plot_color in plot_colors.items():
    time, num_dens[nuclide] =results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
    print(time, num_dens[nuclide])
    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```

anu1217 replied:
Missed this comment -- will add it to the code

gonuke commented on PointSourceALARA:
```diff
@@ -0,0 +1,46 @@
+
+
+Schedule myschedule
+	100 s	  flux_3   single_pulse     0  s
```
Need to update this to match OpenMC

anu1217 committed:
Fixed a typo in one of the bounds

anu1217 committed:
Delete VitJ.txt

- Replacing with .xlsx file due to errors reading data from .txt file

anu1217 committed:
Adding .xlsx with Vitamin-J energy bounds

anu1217 committed:
Binning fluxes according to Vitamin-J

- Also creating .xlsx files to store flux spectrum and tally values

anu1217 committed:
Imported pandas module

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,174 @@
+
+# Importing Vitamin-J energy group structure:
+# This excel file contains the energy bounds of the Vitamin J structure
+Vit_J = pd.read_excel('VitaminJEnergyGroupStructure.xlsx')
```
Why did you switch to XLSX?

anu1217 replied:
It was having trouble reading from the txt file and giving a tally output. For some reason it appears to work fine with the Excel file

gonuke replied:
I'm sure we can figure out how to get the txt file version to work - certainly better than adding proprietary software files.  It may be moot for this particular case given that we will likely switch to the built in Vit-J groups.

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,174 @@
+
+# Creating a tally to get the flux energy spectrum.
+# An energy filter is created to assign to the flux tally using the Vitamin J structure.
+energy_filter_flux = openmc.EnergyFilter(ebounds)
```
I think I just learned that OpenMC has the 175 group Vitamin-J structure built in.

See the `from_group_structure` method [here](https://docs.openmc.org/en/stable/pythonapi/generated/openmc.EnergyFilter.html) and the library of structures [here](https://docs.openmc.org/en/v0.12.1/pythonapi/mgxs.html)

anu1217 replied:
Do you have a preference as to how the structure is implemented? (I'm guessing the built-in one)

gonuke replied:
I think it's best to switch to the built-in groups, yes.

anu1217 replied:
Sounds good. Made this change and the results appear to be the same.

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,174 @@
+with openmc.StatePoint('statepoint.10.h5') as sp:
+    fl = sp.get_tally(name="Flux spectrum")
+    nt = sp.get_tally(name="Neutron tally")
+    # Get the neutron energies from the energy filter
+    energy_filter_fl = fl.filters[0]
+    energies_fl = energy_filter_fl.bins[:, 0]
+
+    # Get the neutron flux values
+    flux = fl.get_values(value='mean').ravel()
+    
+    #Neutron flux/elastic/absorption tallies:
+    tal = nt.get_values(value='mean').ravel()
```
These lines don't have to be inside the `with` block, do they?

anu1217 replied:
I think they do, or else I get an "expected an indented block" error

anu1217 replied:
Corrected indentation

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,174 @@
+    tal = nt.get_values(value='mean').ravel()
+
+Flux_Data = np.c_[energies_fl, flux]
+#Creating an excel file that stores flux data for each energy bin (used as input for ALARA)
```
Not sure why we want Excel?

gonuke replied:
Aside from my personal distaste for Excel as a data processing tool, I don't think it adds anything to make this an Excel file vs some kind of text file

anu1217 committed:
Removed indentation in lines 95-103

anu1217 committed:
Removed stable W isotopes from list of plotted items

anu1217 committed:
Removed stable isotopes from results.get_atoms() call

anu1217 committed:
Plotting nuclide densities on log-log scale

anu1217 committed:
Update PointSourceALARA to match OpenMC input

- Changed operating & cooling times to match those in OpenMC
- Added normalization factor in flux line
- Added some comments about these changes

gonuke commented on PointSourceALARA:
```diff
@@ -19,21 +19,25 @@ mixture W_wall
+# This gives fluxes in #/cm^2-s when multiplied by the values in flux_ss
 
-flux flux_3   fluxdata/flux_ss	1E+18 	0	default
+flux flux_3   fluxdata/flux_ss	15836181434 	0	default
```
why this change?

gonuke replied:
OpenMC is still using 1e18

anu1217 replied:
This is 1e18 normalized by the volume of the material, so that flux is in #/cm^2-s. There was a 6-7 order of magnitude difference in number density between OpenMC and ALARA without this

gonuke replied:
ALARA expects a flux in n/cm^2-s, so it's probably better to divide by the volume before writing the flux so that it's already in n/cm^2/source , rather than dividing manually.  That will help us see that the strengths are the same.

anu1217 committed:
Storing & printing nuclide densities from list

anu1217 committed:
OpenMC now reads Vitamin-J bounds from its built-in structure

- The lines referring to the Excel file which was previously used have been commented out

anu1217 committed:
Plotting all nuclides from depletion results

- Storing + plotting results from all nuclides from activation instead of a select few

gonuke commented on VitaminJEnergyGroupStructure.xlsx:
can we delete this file now?

anu1217 replied:
File has been deleted

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,165 @@
+tal = nt.get_values(value='mean').ravel()
+print(tal)
+
+Flux_Data = np.c_[energies_fl, flux]
```
Here is where you may want to normalize the flux by the volume.  Strictly speaking, it's not a *flux*, per se, until you divide by the volume.

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,165 @@
+num_dens= {}
+pair_list = {}
+
+k = 0
+for k in range(len(nuc_last)) :
+    Random_Color = random.choice(colors)
+    plot_colors = {nuc_last[k]: Random_Color}
+    pair_list.update(plot_colors)
+    k = k + 1
+
+for nuclide, plot_color in pair_list.items():
+    time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+    print(time, num_dens[nuclide])
+    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```
```suggestion
for nuclide in nuc_last:
    plot_color = random.choice(colors)
    time, num_dense[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
    print(time, num_dens[nuclide])
    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```

gonuke reviewed (changes requested):
This is definitely getting closer.  I recommend more modularity overall in the OpenMC script.

anu1217 committed:
Deleting VitaminJ Excel file

anu1217 committed:
Merge branch 'anu1217-openmc-sim' of https://github.com/anu1217/OpenMCActivationStudy into anu1217-openmc-sim

anu1217 committed:
Condensing nuclide density plotting to 1 loop

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Fixed a typo

anu1217 committed:
Changing ALARA flux multiplier back to 1E+18

anu1217 committed:
Dividing OpenMC neutron fluxes by volume to get proper flux units

- Also made fluxes go from high energy to low energy to match ALARA input

anu1217 committed:
Flux and tally values are now saved as .csv instead of .xlsx

anu1217 committed:
Changing a couple of variable names

gonuke commented on myElelib:
Is this the standard element library, or a customized one?  If the former, then we probably don't need it in the repo

anu1217 replied:
It's the library from the sample folder of ALARA. It's different from the standard one in that it has some other isotopes (e.g. Li-90) added in, but this doesn't matter for our W-only problem

gonuke commented on fendl2bin.lib:
Don't need this file in the repo

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,163 @@
+    plot_color = random.choice(colors)
+    time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+    print(time, num_dens[nuclide])
+    plt.plot(time, num_dens[nuclide], marker='.', linestyle='solid', color=plot_color, label=nuclide)
```
I'm noticing that your x-axis (time) on your plot goes up to just over 120,000 but you are depleting through 2.6e6 seconds... is something missing?

gonuke replied:
Related: I don't think the nuclides that remain at this time will experience much decay over 120,000 s.

gonuke replied:
Perhaps plot the y-axis on a log scale?

anu1217 replied:
The graph is already on a log scale, but I was misinterpreting how OpenMC and Python track time.

anu1217 replied:
I think the x-limits should be adjusted and the graph title changed to read "time after operation" rather than shutdown

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,163 @@
+stable_nuc = ['W180', 'W182', 'W183', 'W184', 'W186']  
+
+#Store list of nuclides from last timestep as a Materials object
+materials_last = results.export_to_materials(-1)
```
If you only plot the nuclides from the last time step, you might get relatively stable nuclides and might not see much decay.

anu1217 replied:
The nuclides in the last timestep include the nuclides created at all earlier timesteps.

anu1217 replied:
However, while I was playing around with the time_steps I found out something interesting. With the current operating time of 3e8 s, around 60 or so nuclides are at the last time step. If a small timestep is added at the beginning (e.g. 1000s) with the same neutron source rate, the number of nuclides at the end is around 230.

anu1217 replied:
This seems to suggest that OpenMC isn't tracking the short-lived nuclides that are created near the beginning of operation, unless that short timestep is added in explicitly. Perhaps we should break up the operating time into smaller chunks to account for this

gonuke reviewed (changes requested):
A few more comments while you work on this more.

anu1217 committed:
Deleting myElelib from repo

anu1217 committed:
Deleting myElelib from repo

anu1217 committed:
Saving nuclide densities as .csv file for later result comparison

anu1217 committed:
Changing formatting to save nuclide densities

gonuke commented on SS_Activation_OpenMC.py:
```diff
@@ -0,0 +1,167 @@
+        plot_color = random.choice(colors)
+        time, num_dens[nuclide] = results.get_atoms('1', nuclide, nuc_units = 'atom/cm3')
+        print(time, num_dens[nuclide])
+        density_file.write(''.join(map(str, nuclide)) + ',')
```
Since this is a single nuclide, this join approach is probably overkill
```suggestion
        density_file.write(f'{nuclide},')
```

gonuke reviewed (changes requested):
Similar to the discussion of Eitan's PR yesterday - I wonder if we should focus on getting some of this merged and settled while you tweak the post-processing/comparisons.

Perhaps we finalize the ALARA and OpenMC input parts and separate it from the post-processing part.

anu1217 committed:
Update density_file.write()

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Adjusting x-limits to correctly reflect time progression

anu1217 committed:
Using IndependentOperator instead of CoupledOperator

- Using the IndependentOperator as the neutron source is nearly independent of material composition in the absence of fissile material

anu1217 commented:
> Similar to the discussion of Eitan's PR yesterday - I wonder if we should focus on getting some of this merged and settled while you tweak the post-processing/comparisons.
> 
> Perhaps we finalize the ALARA and OpenMC input parts and separate it from the post-processing part.

I was going to start a new pull request for the comparisons part, so I'm good with merging this if you are

gonuke commented:
I'm fine to merge this now, but we should discuss some code design issues for the next phase.
